[FLINK-14561] Don't write FLINK_PLUGINS_DIR env variable to Configuration

tillrohrmann opened the pull request:
## What is the purpose of the change

This commit reads the FLINK_PLUGINS_DIR env variable directly instead of first
writing it first to the Flink Configuration and then reading it from there. This
has the advantage that the plugins env variable which is intended for the local
process only is not being leaked to other processes. The latter can happen if
we start a new Flink process based on the read Flink Configuration (e.g. Yarn,
Mesos TaskExecutor).

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

wangyang0918 commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -2012,6 +2012,9 @@
 	/** The environment variable name which contains the location of the plugins folder. */
 	public static final String ENV_FLINK_PLUGINS_DIR = "FLINK_PLUGINS_DIR";
 
+	/** The default Flink plugins directory is none has been specified via {@link #ENV_FLINK_PLUGINS_DIR}. */
```
Is it a typo? is -> if
/** The default Flink plugins directory if none has been specified via {@link #ENV_FLINK_PLUGINS_DIR}. */

GJL replied:
It's a typo.

tillrohrmann replied:
Yes it's a typo. Will correct it.

wangyang0918 commented on flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java:
```diff
@@ -55,23 +55,22 @@ private PluginConfig(Optional<Path> pluginsPath, String[] alwaysParentFirstPatte
+
 		File pluginsDirFile = new File(pluginsDir);
 		if (!pluginsDirFile.isDirectory()) {
 			LOG.warn("Environment variable [{}] is set to [{}] but the directory doesn't exist",
```
If the env is not set and DEFAULT_FLINK_PLUGINS_DIRS does not exist, it will show a confusing log. 
`Environment variable [FLINK_PLUGINS_DIR] is set to [plugins] but the directory doesn't exist`

tillrohrmann replied:
True. Let me change it into "The plugins directory [plugins] does not exist."

wangyang0918 commented:
Failed test `testConfigureWithMissingPlugins` need to take a look.

GJL commented on flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java:
```diff
@@ -55,23 +55,22 @@ private PluginConfig(Optional<Path> pluginsPath, String[] alwaysParentFirstPatte
-			LOG.info("Environment variable [{}] is not set", ConfigConstants.ENV_FLINK_PLUGINS_DIR);
-			return Optional.empty();
-		}
+	public static Optional<File> getPluginsDir() {
```
nit: `Path` is considered the more modern API.

https://stackoverflow.com/questions/6903335/java-7-path-vs-file

The code below can be rewritten to:

```
final Path pluginsDirFile = Paths.get(pluginsDir);
if (!Files.isDirectory(pluginsDirFile)) {
```

Other places would need to be rewritten, however. Therefore, I am not insisting on a change.

tillrohrmann replied:
I deliberately used `File` in order to express that the directory needs to be local to the process. Happy to change it at a later point when we decide that the plugins directory can reside on any filesystem.

GJL replied:
Ok I am fine with keeping the code. However, I don't think that `java.nio.file.Path` conveys that the file can be potentially on a remote filesystem. On my platform there are only local `java.nio.file.FileSystem` implementations.

tillrohrmann replied:
True, but this must not hold for every other setup.

GJL commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/overlays/FlinkDistributionOverlay.java:
```diff
@@ -57,9 +59,9 @@
+	final Optional<File> flinkPluginsPath;
 
-	public FlinkDistributionOverlay(File flinkBinPath, File flinkConfPath, File flinkLibPath, File flinkPluginsPath) {
+	public FlinkDistributionOverlay(File flinkBinPath, File flinkConfPath, File flinkLibPath, Optional<File> flinkPluginsPath) {
```
https://flink.apache.org/contributing/code-style-and-quality-java.html#java-optional

>    Do not use Optional as a function argument, instead either overload the method or use the Builder pattern for the set of function arguments.
>        Note: an Optional argument can be allowed in a private helper method if you believe that it simplifies the >code (example).
 >   Do not use Optional for class fields.

tillrohrmann replied:
If you insist, then I'll change it to a `@Nullable` parameter. But it won't change much imo.

GJL replied:
I think that 

> Do not use Optional for class fields.

is a clear statement, and we should respect it because it causes a warning in some setups.

GJL replied:
Constructor of FlinkDistributionOverlay can be even made private.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/overlays/FlinkDistributionOverlay.java:
```diff
@@ -75,12 +77,7 @@ public void configure(ContainerSpecification container) throws IOException {
-		else {
-			LOG.warn("The plugins directory '" + flinkPluginsPath + "' doesn't exist.");
-		}
+		flinkPluginsPath.ifPresent(FunctionUtils.uncheckedConsumer(pluginsPath -> addPathRecursively(pluginsPath, TARGET_ROOT, container)));
```
The check has been moved to `PluginConfig#getPluginsDir`. Now the assumption is that `FlinkDistributionOverlay` must be initialized with a existing plugins directory if `flinkPluginsPath` is not empty.

buptljy replied:
LGTM.

tillrohrmann commented:
Thanks for your review @GJL, @wangyang0918 and @buptljy. I've addressed most of your comments and answered those comments with which I did not strictly agree. Please take another look.

wangyang0918 commented:
LGTM, ready for merge.

tillrohrmann commented:
Thanks a lot for the review @wangyang0918, @GJL and @buptljy. I've rebased and merging this PR once Travis gives green light.
